Add FuseCrossJoinedGlobalAggregations rule

lukasz-stec opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Add a simplified version of the `JoinOnKeys` rule
from "Computation Reuse via Fusion in Amazon Athena" paper that acts only on cross join over global
aggregations over table scans with possible filter. This can transform:
```
select from (select count(*) c1 from table where a = 1),
  (select count(*) c2 from table where b = 1)
```
into
```
select count(*) filter (where a = 1),
 count(*) filter (where b = 1)
from table where a = 1 or b = 1
```
making the query read the table only once.

<!-- Provide a user-friendly explanation, keep it brief if it isn't user-visible. -->
## Non-technical explanation

Speed up queries that have a subquery that matches the specific pattern of
cross join over global aggregation on the same table by reading the table only once.

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( X) Release notes are required, with the following suggested text:

```markdown
# General
* Speed up a cross, self joins over aggregations by reading the table only once
```

lukasz-stec commented:
Improvement for a basic query that matches this rule:
before
```
explain SELECT * from (SELECT count(*) c_1 FROM lineitem WHERE suppkey = 3332274), (SELECT count(*) c_2 FROM lineitem WHERE suppkey = 1296500);

trino:tpch_sf1000_dec_orc_part> SELECT * from (SELECT count(*) c_1 FROM lineitem WHERE suppkey = 3332274), (SELECT count(*) c_2 FROM lineitem WHERE suppkey = 1296500);
 c_1 | c_2 
-----+-----
 783 | 485 
(1 row)

Query 20220926_104253_00011_p5eka, FINISHED, 6 nodes
http://localhost:8081/ui/query.html?20220926_104253_00011_p5eka
Splits: 5,378 total, 5,378 done (100.00%)
CPU Time: 473.1s total, 25.4M rows/s, 86.3MB/s, 19% active
Per Node: 10.1 parallelism,  255M rows/s,  867MB/s
Parallelism: 60.3
Peak Memory: 1.75KB
7.84 [12B rows, 39.9GB] [1.53B rows/s, 5.08GB/s]
```
after
```
trino:tpch_sf1000_dec_orc_part> set session fuse_sub_plan=true;
SET SESSION
trino:tpch_sf1000_dec_orc_part> SELECT * from (SELECT count(*) c_1 FROM lineitem WHERE suppkey = 3332274), (SELECT count(*) c_2 FROM lineitem WHERE suppkey = 1296500);
 c_1 | c_2 
-----+-----
 783 | 485 
(1 row)

Query 20220926_104318_00013_p5eka, FINISHED, 6 nodes
http://localhost:8081/ui/query.html?20220926_104318_00013_p5eka
Splits: 2,674 total, 2,674 done (100.00%)
CPU Time: 255.4s total, 23.5M rows/s, 79.8MB/s, 23% active
Per Node: 9.5 parallelism,  224M rows/s,  760MB/s
Parallelism: 57.1
Peak Memory: 6.75KB
4.47 [6B rows, 19.9GB] [1.34B rows/s, 4.45GB/s]
```

gaurav8297 commented on core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java:
```diff
@@ -756,4 +758,17 @@ public OptimizerConfig setUseExactPartitioning(boolean useExactPartitioning)
+        return fuseSubPlan;
+    }
+
+    @Config("optimizer.fuse-sub-plan")
```
Isn't this should be more descriptive? As in, right now we only support a particular kind of fusing.

lukasz-stec replied:
IMO it's better to have one toggle for all fusing rules that are incoming.

gaurav8297 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantIdentityProjections.java:
```diff
@@ -33,7 +33,7 @@
             .matching(RemoveRedundantIdentityProjections::outputsSameAsSource);
 
-    private static boolean outputsSameAsSource(ProjectNode node)
+    public static boolean outputsSameAsSource(ProjectNode node)
```
Instead of making it public, I think it's better to have this method inside `FuseCrossJoinedGlobalAggregations`

lukasz-stec replied:
then I would have to copy it to `PlanNodeFuser` as well.

gaurav8297 replied:
I think it's fine to have it inside `PlanNodeFuser` which can be used by `FuseCrossJoinedGlobalAggregations`. It's just that it might not be a good idea to depend on other rule public methods.

lukasz-stec replied:
Ok, i copied it to the `PlanNodeFuser`

gaurav8297 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,143 @@
+                        result = new FilterNode(
+                                context.getIdAllocator().getNextId(),
+                                fused.plan(),
+                                and(fused.leftFilter(), fused.rightFilter()));
```
Isn't this have to be `or`?

lukasz-stec replied:
Not according to the paper

> JoinOnKeys
...
An interesting special case of
this rule happens when both K1 and K2 are empty (e.g., for
scalar aggregates) and the subsequent join is a cross product. In
this case, given GroupBy,A1(Q1)⨯GroupBy,A2(Q2) and
Fuse(Q1, Q2) = (Q, M, L, R), we transform the input
pattern into FilterL AND R(GroupBy,A1 M(A2)(Q)).

That said, i don't think I have it tested and I'm not even sure I can setup this case using current fuse. 
I will double check

gaurav8297 replied:
I think both left and right filters will always be `true` if you are not dealing with the non-scaler GroupBy

lukasz-stec replied:
yeah, currently this always gonna be true because fused aggregations do not expose filters (those are applied inside the aggregation using a mask)
I dropped the if and added precondition

gaurav8297 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,433 @@
+ * Implements simplified version of fusion operation from the
+ * "Computation Reuse via Fusion in Amazon Athena"  paper.
+ */
+public class PlanNodeFuser
```
Can we add more comments to this class? Basically how each method works with an example?

lukasz-stec replied:
I didn't want to copy the paper(it's all there)  but I guess it's better to the reader to have it here. Will do.

lukasz-stec replied:
comments added (based on the paper)

gaurav8297 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,433 @@
+        }
+
+        PlanNode fused;
+        if (leftPredicate.equals(TRUE_LITERAL) || mappedRightPredicate.equals(TRUE_LITERAL)) {
```
I think instead of `||` we should have `&` here. Let me know if I'm missing something.

gaurav8297 replied:
Basically what if right predicate isn't TRUE while left is?

lukasz-stec replied:
Fusing `Filter` basically pushes the left and right conditions to the "fused" conditions but it also has to add `left or right` to the plan to filter out rows that do not match any side.
If either left or right condition is `TRUE` then `left or right` also equals `TRUE` so we don't have to add the filter.
This means basically that one side is full scan and so we need to push all rows anyway.
Hope that helps. Do you think it's with extending the comment there?

gaurav8297 replied:
This makes sense. Yeah I think we can add a small comment here

lukasz-stec replied:
comment extended

gaurav8297 commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/fuse/TestFuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,503 @@
+    }
+
+    @Test
+    public void testTableScanWithFilterOnlyOnOneTheRIghtSide()
```
`testTableScanWithFilterOnlyOnOneTheRIghtSide` -> `testTableScanWithFilterOnlyOnTheRightSide`

gaurav8297 reviewed (commented):
Some initial comments

lukasz-stec reviewed (commented):
most comments addressed, the documentation of the fuse operation is still pending

gaurav8297 commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/fuse/TestFuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,503 @@
+                                        tableScan(TEST_TABLE_NAME, ImmutableMap.of("colLeft", "col")))));
+    }
+
+    // this fails due to test infra assumption that projection always has unique target expression
```
Can we somehow fix this? Because it seems like a pretty imp test

lukasz-stec replied:
fixing the entire infra is hard and I don't know if we want to do it just for this case because it makes a lot of things easier.
Also, I have this case tested in `AbstractTestJoinQueries.testFuseCrossJoinOnGlobalAggregationWithDuplicatedAggregation`.

gaurav8297 commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/fuse/TestFuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,503 @@
+import static io.trino.testing.TestingHandles.TEST_CATALOG_HANDLE;
+import static io.trino.testing.TestingHandles.TEST_CATALOG_NAME;
+
+public class TestFuseCrossJoinedGlobalAggregations
```
Can you add a test where both table scan outputs two different columns from the same table and then aggregation is applied separately on these columns?

```
tableScanA -> p.tableScan(TEST_TABLE_HANDLE, List.of(colLeft), ImmutableMap.of(colLeft, TEST_TABLE_COLUMN_A)

tableScanB -> p.tableScan(TEST_TABLE_HANDLE, List.of(colLeft), ImmutableMap.of(colLeft, TEST_TABLE_COLUMN_B)
```

lukasz-stec replied:
added `testDifferentColumns`

gaurav8297 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,138 @@
+        PlanNode left = context.getLookup().resolve(node.getLeft());
+        PlanNode right = context.getLookup().resolve(node.getRight());
+
+        while (left instanceof ProjectNode && outputsSameAsSource((ProjectNode) left)) {
```
I think these conditions are not required if this rule is fired after `RemoveRedundantIdentityProjections`

lukasz-stec replied:
there is an order of rule issue. this rule cannot be run after `RemoveRedundantIdentityProjections` because it has to be run before columnPruningOptimizer, PushPredicateIntoTableScan, AddExchanges.
I could add another `RemoveRedundantIdentityProjections` before this rule but this changes plans for some unrelated queries on values and I want this change to be as small as possible.

gaurav8297 replied:
I don't there's any harm in adding `RemoveRedundantIdentityProjections` and it simplifies the fuse rule since it only has to take care of the fusing logic and nothing else. WDYT?

lukasz-stec replied:
There are two things. One is, adding this `RemoveRedundantIdentityProjections` before `FuseCrossJoinedGlobalAggregations` changes plans in unrelated queries in a not-obvious way (it reorders the symbols in `VALUES` node) but also it runs `RemoveRedundantIdentityProjections` for the entire plan even if the `FuseCrossJoinedGlobalAggregations` won't fire, adding to the planning time.

gaurav8297 commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,138 @@
+            right = context.getLookup().resolve(((ProjectNode) right).getSource());
+        }
+
+        if (!(right instanceof AggregationNode) || !((AggregationNode) right).getGroupingKeys().isEmpty()) {
```
Can we have these conditions as part of pattern matching for this rule?

lukasz-stec replied:
not at the moment because of the identity projection pruning

gaurav8297 commented on core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/fuse/TestFuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,503 @@
+import static io.trino.testing.TestingHandles.TEST_CATALOG_HANDLE;
+import static io.trino.testing.TestingHandles.TEST_CATALOG_NAME;
+
+public class TestFuseCrossJoinedGlobalAggregations
```
Do you think we should also handle this case which seems trivial? Basically when the projection is almost identical

```
p.project(Assignments.identity(colA),
          p.tableScan(TEST_TABLE_HANDLE, List.of(colA, colB), ImmutableMap.of(colA, TEST_TABLE_COLUMN_A, colB, TEST_TABLE_COLUMN_B)))
```

lukasz-stec replied:
It's easy to add full projection support and I plan to do it in the follow-up PRs. I want to make small, simple steps.
The reason we are supporting full identify projections now is that they are there in the plan even for the simplest query.

gaurav8297 reviewed (commented):
Please rebase

raunaqmorarka reviewed (commented):
preliminary comments, will add more

lukasz-stec commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,138 @@
+
+        while (left instanceof ProjectNode && outputsSameAsSource((ProjectNode) left)) {
+            // ignore identity projection
+            left = context.getLookup().resolve(((ProjectNode) left).getSource());
```
how this is different from `resolve`?

lukasz-stec replied:
this was dropped from here (I'm using captures). There is still resolve in the fuser.

lukasz-stec commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,138 @@
+            // ignore identity projection
+            left = context.getLookup().resolve(((ProjectNode) left).getSource());
+        }
+        while (right instanceof ProjectNode && outputsSameAsSource((ProjectNode) right)) {
+            // ignore identity projection
+            right = context.getLookup().resolve(((ProjectNode) right).getSource());
+        }
```
Adding has unintended consequences, see https://github.com/trinodb/trino/pull/14271#discussion_r988976782.
In order to keep this change small and self-contained IMO it's better to ignore identity projections here.

lukasz-stec replied:
after offline discussion, I used `RemoveRedundantIdentityProjections` (this forced some changes in the unrelated plan tests)

lukasz-stec commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,138 @@
+            return Result.empty();
+        }
+
+        if (!(left instanceof AggregationNode) || !((AggregationNode) left).getGroupingKeys().isEmpty()) {
```
> `private static boolean isGlobalAggregation(PlanNode node)...`

why?

> If I'm understanding correctly, we're requiring that this rule runs before `PushPartialAggregationThroughExchange` creates partial aggregations ?

Well yes, but even more, this has to run before `AddExchanges` to not deal with exchanges here.

lukasz-stec replied:
used `AggregationNode.isSingleGlobalAggregation`

lukasz-stec commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,138 @@
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join()
+            .matching(JoinNode::isCrossJoin);
```
Do you mean instead of 
```
PlanNode left = context.getLookup().resolve(node.getLeft());
PlanNode right = context.getLookup().resolve(node.getRight());
``` 
?

lukasz-stec replied:
done

lukasz-stec commented:
added prefix PR https://github.com/trinodb/trino/pull/14559 for `AggregationNode.isSingleGlobalAggregation`. Only the last commit matters here.
Also, addressed all comments except for docs

lukasz-stec commented:
@gaurav8297 @raunaqmorarka FYI this is ready for review

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,530 @@
+            return fuseFilter(left, TRUE_LITERAL, rightFilter.getSource(), rightFilter.getPredicate());
+        }
+
+        throw new IllegalArgumentException(String.format("expected at least one FilterNode but got %s and %s", left, right));
```
nit: static import `format`

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,530 @@
+        Set<Symbol> rightOutputSymbols = ImmutableSet.copyOf(right.getOutputSymbols());
+
+        // add unique assignments and outputs from the right side or map to the left side symbols
+        right.getAssignments().forEach(((symbol, columnHandle) ->
```
symbol -> rightSymbol

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,530 @@
+                                entry -> {
+                                    // left assignment found, add the mapping and output symbol if necessary
+                                    symbolMappingBuilder.add(symbol, entry.getKey());
+                                    if (rightOutputSymbols.contains(symbol)) {
```
Isn't this always true since we require that `assignments.keySet().containsAll(outputs)` in table scan node ?

raunaqmorarka replied:
I'm thinking why don't we loop on `rightOutputSymbols` instead and only lookup the assignments which are part of `rightOutputSymbols` ? I don't think we need to care about the assignments which are not part of output symbols as `PruneTableScanColumns` would've gotten rid of them anyway.

raunaqmorarka replied:
ping

lukasz-stec replied:
why does `TableScanNode` have separate output symbols if they are not needed?
I feel like prunning the assignments here is not the right place because this rule shouldn't care if assignments are needed or not.

lukasz-stec commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,530 @@
+                        assignments,
+                        left.getEnforcedConstraint(),
+                        left.getStatistics(),
+                        left.isUpdateTarget(),
```
since we are requiring both `isUpdateTarget()` are equal, why not leave it as is? this way the code is independent of the "can't be 2 table scans which are update target in the plan" rule.

lukasz-stec commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,530 @@
+                        left.getEnforcedConstraint(),
+                        left.getStatistics(),
+                        left.isUpdateTarget(),
+                        left.getUseConnectorNodePartitioning()),
```
hmm, I think `Optional.empty` would break the planner if the rule is run after `DetermineTableScanNodePartitioning`

raunaqmorarka replied:
If we were running this rule after DetermineTableScanNodePartitioning then just setting useConnectorNodePartitioning to empty and adding DetermineTableScanNodePartitioning to the optimizer loop would be enough to handle it

lukasz-stec replied:
I added `checkState(left.getUseConnectorNodePartitioning().equals(right.getUseConnectorNodePartitioning()),...`, 
using `left.getUseConnectorNodePartitioning()` makes this rule more order independent.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,530 @@
+                        ImmutableList.copyOf(outputs),
+                        assignments,
+                        left.getEnforcedConstraint(),
+                        left.getStatistics(),
```
That link didn't work for me.
From what I can tell, since we are doing this before predicate pushdown, `outputRowCount` should be the same for both scans, so it should be the same for the fused scan as well. When adding symbols to `outputs` above, we can also lookup the symbol stats for that from whichever TS it is part of and use that to build symbol stats of the fused scan.

lukasz-stec replied:
> That link didn't work for me.

this should work. https://github.com/trinodb/trino/pull/14271#discussion_r1009257047

Instead of merging stats hereI added preconditions that stats are empty.

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,530 @@
+        }
+
+        if (!left.getEnforcedConstraint().equals(right.getEnforcedConstraint()) ||
+                !left.getStatistics().equals(right.getStatistics()) ||
```
I think it should be ok to set the stats for the fused TS to empty and let subsequent rules derive stats, we did similar thing in ApplyTableScanRedirection. Then this check can be removed

lukasz-stec reviewed (commented):
CA

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,542 @@
+        });
+        return Optional.of(new FusedPlanNode(
+                new TableScanNode(
+                        context.getIdAllocator().getNextId(),
```
`left.getId`
We should be able to re-use `id` because if fuse is successful then the original `TableScanNode` would be discarded

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,542 @@
+
+        Expression mappedRightPredicate = fusedSource.symbolMapping().map(rightPredicate);
+        if (leftPredicate.equals(mappedRightPredicate)) {
+            // simplified version. since the predicates match, and are applied in the fused FilterNode as predicate,
```
Remove `simplified version.`

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,542 @@
+
+/**
+ * Implements simplified version of fusion operation from the
+ * "Computation Reuse via Fusion in Amazon Athena"  paper.
```
extra space in `Athena"  paper.`

lukasz-stec commented on testing/trino-tests/src/test/resources/sql/presto/tpcds/hive/partitioned/q66.plan.txt:
```diff
@@ -1,8 +1,8 @@
-        final aggregation over (ship_carriers, w_city, w_country, w_county, w_state, w_warehouse_name, w_warehouse_sq_ft, year)
-            local exchange (REPARTITION, HASH, ["ship_carriers", "w_city", "w_country", "w_county", "w_state", "w_warehouse_name", "w_warehouse_sq_ft", "year"])
-                partial aggregation over (concat_337, d_year, w_city_2, w_country_5, w_county_3, w_state_4, w_warehouse_name_0, w_warehouse_sq_ft_1)
+        final aggregation over (ship_carriers$gid, w_city$gid_208, w_country$gid_211, w_county$gid_209, w_state$gid_210, w_warehouse_name$gid_206, w_warehouse_sq_ft$gid_207, year$gid)
```
yes, that's correct. Only adding `RemoveRedundantIdentityProjections` (without `FuseCrossJoinedGlobalAggregations`) has the same effect plus `FuseCrossJoinedGlobalAggregations` does not fire for any tpcds/tph query

lukasz-stec commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,542 @@
+
+        public Symbol map(Symbol symbol)
+        {
+            Symbol mapped = mapping.get(symbol);
+            return mapped != null ? mapped : symbol;
```
The logic is that we want to use the mapped (`left`) symbol if the mapping exists and use the `right` symbol if it does not. This stems directly from the fuse operation definition (see below) that either the `right` symbol is unique (not on the `left`) or it should be mapped. Put it differently, if there is no mapping than the `right` symbol has to be in the fused plan outputs.
```
If Fuse(P1, P2) = (P, M, L, R), then:
- P is the fused resulting plan. The schema of P includes
all output columns in P1 and, optionally, additional
output columns from P2.
- M is a mapping from the output columns of P2 to output
columns of P.
- L and R are two filter conditions defined over the output
columns of P to restore P1 and P2, respectively
```

> What if we add identity symbol mappings for the symbols from left node output

that wouldn't work. We could add identity mappings from the _right_ node but that would be deviation from the paper (the paper describes in detail how the mapping is constructed for each plan node) and IMO keeping the implementation close to the paper makes it more straightforward and less error prune

lukasz-stec commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,542 @@
+        public FusedSymbolMapping(Map<Symbol, Symbol> mapping)
+        {
+            this.mapping = requireNonNull(mapping, "mapping is null");
+            this.symbolMapper = new SymbolMapper(this::map);
```
We could, but I think it's useful. First, it directly matches the paper definition, second, it allows for convenience methods like `map` (with default return) and `requiredMap`

lukasz-stec reviewed (commented):
CA

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,539 @@
+        for (Map.Entry<Symbol, Aggregation> aggregationEntry : right.getAggregations().entrySet()) {
+            Aggregation aggregation = aggregationEntry.getValue();
+
+            Expression newMask = null;
```
Optional instead of Nullable for consistency

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,539 @@
+                    aggregationMask = Optional.of(newMaskSymbol);
+                }
+
+                Aggregation mappedAggregation = new Aggregation(
```
Inline into next line

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java:
```diff
@@ -756,4 +758,17 @@ public OptimizerConfig setUseExactPartitioning(boolean useExactPartitioning)
+        return fuseSubPlan;
+    }
+
+    @Config("optimizer.fuse-sub-plan")
```
Maybe `optimizer.combine-similar-sub-plans`
Fuse or fusion is not a common term

lukasz-stec commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,539 @@
+                Expression newMask = aggregation.getMask()
+                        .map(filter -> and(filter.toSymbolReference(), fusedSource.leftFilter()))
+                        .orElse(fusedSource.leftFilter());
+                Symbol newMaskSymbol = context.getSymbolAllocator().newSymbol("aggr_mask", BOOLEAN);
```
is that a good idea given that the new mask is different? Is it possible that other expressions use the old mask?

lukasz-stec reviewed (commented):
CA

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,534 @@
+                TRUE_LITERAL));
+    }
+
+    private static boolean maskMatches(Optional<Symbol> optionalMaskSymbol, Map<Symbol, Expression> masks, Optional<Symbol> otherMaskSymbol, Optional<Expression> otherMask)
```
Can we simplify this to
```
private static boolean maskMatches(Optional<Symbol> maskSymbol, Map<Symbol, Expression> masks, Optional<Symbol> otherMaskSymbol, Optional<Expression> otherMask)
    {
        Optional<Expression> mask = maskSymbol.map(masks::get);
        if (mask.isEmpty()) {
            // no fused mask and symbols must match
            return otherMask.isEmpty() && maskSymbol.equals(otherMaskSymbol);
        }
        return mask.equals(otherMask);
    }
```
Current code looks a bit complicated

raunaqmorarka commented:
Please rebase to fix conflicts

lukasz-stec commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,534 @@
+                        leftAggregation.isDistinct() == rightAggregation.isDistinct() &&
+                        leftAggregation.getOrderingScheme().equals(rightAggregation.getOrderingScheme()) &&
+                        leftAggregation.getFilter().equals(mappedFilter) &&
+                        maskMatches(leftAggregation.getMask(), additionalAggregationMasks, rightAggregation.getMask(), newMask)) {
```
Good catch, I think `mappedMask` should be used. I need to figure out a test case for this (in progress)

lukasz-stec reviewed (commented):
comments addressed, test pending for the aggregation with mask case

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,135 @@
+                    }
+                    if (!additionalProjections.isEmpty()) {
+                        result = new ProjectNode(
+                                context.getIdAllocator().getNextId(),
```
We can probably use `node.getId()` here as the original JoinNode should get discarded

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java:
```diff
@@ -0,0 +1,135 @@
+        AggregationNode right = captures.get(RIGHT_AGGREGATION_NODE);
+
+        return new PlanNodeFuser(context).fuse(left, right)
+                .map(fused -> {
```
Okay, are the filters going to be handled differently based on what is the parent node ?
Can we move the part about adding projections into PlanNodeFuser by adding `Set<Symbol> resultOutputSymbols` to API or is this part also specific to this rule ?

lukasz-stec reviewed (commented):
CA

raunaqmorarka commented on core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/PlanNodeFuser.java:
```diff
@@ -0,0 +1,557 @@
+{
+    private final Rule.Context context;
+
+    public PlanNodeFuser(Rule.Context context)
```
I think this doesn't need to be public anymore
